### PR TITLE
Adding GenerateSourceWithPublicTypesAsInternal task

### PR DESCRIPTION
**Pull request by joperezr**
cc: @ericstj @ahsonkhan @safern 

Adding Task that will generate a copy of the specified files while rewriting all public types to internal so that they can be packaged as a source package. We will use this as part of creating our System.Text.Json source package.

NOTE: This can still be considered as WIP since I need to add the targets that we will call from the build here as well, so that will be part of this PR.

### CODE REVIEW

File: src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateSourceWithPublicTypesAsInternal.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+                string sourcePath = file.GetMetadata("SourcePath");
+                string destinationPath = file.GetMetadata("DestinationPath");
+                if (string.IsNullOrEmpty(sourcePath) || string.IsNullOrEmpty(destinationPath))
+                {
+                    Log.LogError($"Item {file.GetMetadata("Identity")} doesn't define SourcePath or DestinationPath metdata.");
+                }
+                _sourceFiles.Add(sourcePath, destinationPath);
+            }
+
+            var sourceTrees = GetSourceTrees();
```
Comment (ahsonkhan) [inline src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateSourceWithPublicTypesAsInternal.cs]: nit: don't use var

File: src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateSourceWithPublicTypesAsInternal.cs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                File.WriteAllText(sourceTree.FilePath, newRootNode.ToFullString());
+            }
+
+            return !Log.HasLoggedErrors;
+        }
+
+        private IEnumerable<SyntaxTree> GetSourceTrees()
+        {
+            List<SyntaxTree> result = new List<SyntaxTree>();
+            foreach (var sourceFile in _sourceFiles)
```
Comment (ahsonkhan) [inline src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateSourceWithPublicTypesAsInternal.cs]: same nit here: don't use var.

File: src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateSourceWithPublicTypesAsInternal.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            var sourceTrees = GetSourceTrees();
+            foreach (SyntaxTree sourceTree in sourceTrees)
+            {
+                PublicTypesToInternalRewriter rewriter = new PublicTypesToInternalRewriter();
+                SyntaxNode newRootNode = rewriter.Visit(sourceTree.GetRoot());
+
+                File.WriteAllText(sourceTree.FilePath, newRootNode.ToFullString());
+            }
+
+            return !Log.HasLoggedErrors;
```
Comment (ahsonkhan) [inline src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateSourceWithPublicTypesAsInternal.cs]: Should we consider an empty `_sourceFiles` dictionary as error?

File: src/Microsoft.DotNet.Build.Tasks.Packaging/src/PublicTypesToInternalRewriter.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+        public override SyntaxNode VisitClassDeclaration(ClassDeclarationSyntax node)
+        {
+            return base.VisitClassDeclaration((ClassDeclarationSyntax)ChangePublicDeclarationToInternal(node));
+        }
+
+        private SyntaxNode ChangePublicDeclarationToInternal(BaseTypeDeclarationSyntax node)
+        {
+            SyntaxToken publicToken;
+            if (HasPublicModifier(node, out publicToken))
```
Comment (ahsonkhan) [inline src/Microsoft.DotNet.Build.Tasks.Packaging/src/PublicTypesToInternalRewriter.cs]: nit: inline `publicToken` declaration

File: src/Microsoft.DotNet.Build.Tasks.Packaging/src/PublicTypesToInternalRewriter.cs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+            {
+                return node.ReplaceToken(publicToken, SyntaxFactory.Token(SyntaxKind.InternalKeyword)
+                                                                    .WithLeadingTrivia(publicToken.LeadingTrivia)
```
Comment (ahsonkhan) [inline src/Microsoft.DotNet.Build.Tasks.Packaging/src/PublicTypesToInternalRewriter.cs]: Why is this necessary? What's an example of leading/lagging "trivia" that we would want to keep?

File: src/Microsoft.DotNet.Build.Tasks.Packaging/src/PublicTypesToInternalRewriter.cs [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+            }
+            else
+            {
+                return node;
+            }
+        }
+
+        private bool HasPublicModifier(BaseTypeDeclarationSyntax token, out SyntaxToken publicToken)
+        {
+            foreach (var modifier in token.Modifiers)
```
Comment (ahsonkhan) [inline src/Microsoft.DotNet.Build.Tasks.Packaging/src/PublicTypesToInternalRewriter.cs]: same nit

File: src/Microsoft.DotNet.Build.Tasks.Packaging/src/PublicTypesToInternalRewriter.cs [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+            {
+                if (modifier.Text.ToLowerInvariant() == "public")
```
Comment (ahsonkhan) [inline src/Microsoft.DotNet.Build.Tasks.Packaging/src/PublicTypesToInternalRewriter.cs]: Is the to lower necessary?

File: src/Microsoft.DotNet.Build.Tasks.Packaging/src/PublicTypesToInternalRewriter.cs [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+                {
+                    publicToken = modifier;
+                    return true;
+                }
+            }
+            publicToken = default(SyntaxToken);
```
Comment (ahsonkhan) [inline src/Microsoft.DotNet.Build.Tasks.Packaging/src/PublicTypesToInternalRewriter.cs]: This won't work?
```C#
publicToken = default;
```

We should update `LangVersion` to latest, if possible.

**Comment by ahsonkhan** [comment]
Were you able to find a solution for choosing and stripping out `#ifdefs` in the sources?

File: src/Microsoft.DotNet.Build.Tasks.Packaging/src/PublicTypesToInternalRewriter.cs (hunk 5 above)
Comment (joperezr) [reply src/Microsoft.DotNet.Build.Tasks.Packaging/src/PublicTypesToInternalRewriter.cs]: These are the whitespaces leading the syntax element and trailing it. If we don't preserve this, then it will replace `    public static class Foo` with `internalstatic class Foo`

File: src/Microsoft.DotNet.Build.Tasks.Packaging/src/PublicTypesToInternalRewriter.cs (hunk 7 above)
Comment (joperezr) [reply src/Microsoft.DotNet.Build.Tasks.Packaging/src/PublicTypesToInternalRewriter.cs]: technically it is not required, so I'll remove it.

**Comment by joperezr** [comment]
> Were you able to find a solution for choosing and stripping out #ifdefs in the sources?

Nope, looks like in order to get that working we are going to have to compile the actual code we are producing (as opposed to just parsing it which what we are doing now)

File: src/Microsoft.DotNet.Build.Tasks.Packaging/src/PublicTypesToInternalRewriter.cs (hunk 8 above)
Comment (joperezr) [reply src/Microsoft.DotNet.Build.Tasks.Packaging/src/PublicTypesToInternalRewriter.cs]: I don't want to require latest version on this whole project(which is very large) just because of one line, so if that's ok, I think I'll leave it as is for now.

**Comment by joperezr** [comment]
@safern do you mind taking a peek at acea945 which contains the targets that will use this task?

File: src/Microsoft.DotNet.Build.Tasks.Packaging/src/PublicTypesToInternalRewriter.cs (hunk 8 above)
Comment (ahsonkhan) [reply src/Microsoft.DotNet.Build.Tasks.Packaging/src/PublicTypesToInternalRewriter.cs]: That's fine by me. Out of curiosity, what's the harm in targeting latest language version, regardless?

Comment (joperezr) [reply src/Microsoft.DotNet.Build.Tasks.Packaging/src/PublicTypesToInternalRewriter.cs]: That might mean that we won't work on older VS versions (which might not be an issue here but I'm not 100% sure)

File: src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateSourceWithPublicTypesAsInternal.cs [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+                Log.LogError("The Files Item cannot be empty.");
+                return false;
+            }
+            foreach (ITaskItem file in Files)
+            {
+                string sourcePath = file.GetMetadata("SourcePath");
+                string destinationPath = file.GetMetadata("DestinationPath");
+                if (string.IsNullOrEmpty(sourcePath) || string.IsNullOrEmpty(destinationPath))
+                {
+                    Log.LogError($"Item {file.GetMetadata("Identity")} doesn't define SourcePath or DestinationPath metadata.");
```
Comment (ahsonkhan) [inline src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateSourceWithPublicTypesAsInternal.cs]: What happens if `file.GetMetadata("Identity")` returns an empty string? Do we have any other meaningful data that we can log or is this unlikely to happen?

File: src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateSourceWithPublicTypesAsInternal.cs [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+                }
+                sourceFiles.Add(sourcePath, destinationPath);
+            }
+
+            IEnumerable<SyntaxTree> sourceTrees = GetSourceTrees(sourceFiles);
+            foreach (SyntaxTree sourceTree in sourceTrees)
+            {
+                PublicTypesToInternalRewriter rewriter = new PublicTypesToInternalRewriter();
+                SyntaxNode newRootNode = rewriter.Visit(sourceTree.GetRoot());
+                if (!Directory.Exists(Path.GetDirectoryName(sourceTree.FilePath)))
```
Comment (ahsonkhan) [inline src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateSourceWithPublicTypesAsInternal.cs]: Cache `Path.GetDirectoryName(sourceTree.FilePath)` in a local?

File: src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateSourceWithPublicTypesAsInternal.cs [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+                {
+                    Directory.CreateDirectory(Path.GetDirectoryName(sourceTree.FilePath));
+                }
+                File.WriteAllText(sourceTree.FilePath, newRootNode.ToFullString());
+            }
+
+            return !Log.HasLoggedErrors;
```
Comment (ahsonkhan) [inline src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateSourceWithPublicTypesAsInternal.cs]: Wouldn't this always be true since whenever we log an error, we return false right after.

Comment (danmosemsft) [reply src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateSourceWithPublicTypesAsInternal.cs]: As a good practice I recommend that all MSBuild task `Execute()` implementations always return only ` !Log.HasLoggedErrors`. That forces you to keep errors and success in sync. MSBuild should have originally made Execute() return void, and done this under the covers.
